chainer/training/extensions/snapshots/writer.py

```diff
@@ -0,0 +1,223 @@
+
+
+class ProcessWriter(Writer):
+    """Writer that uses a process.
+
+    This class creates a process that invokes the actual saving function.
+    A new process is created every time `__call__` is invoked.
+
+    .. note::
+        Forking or spawing a new process from MPI process might be danger.
+        Consider using `ThreadWriter` instead of `ProcessWriter` if you are
+        using MPI in ChainerMN.
+
+    Args:
+        savefun: Callable object. Must have same interface to `Writer.__call__`
+            method. Also must need to be able to passed to a thread.
+    """
+
+    def __init__(self, savefun=util.save_npz, **kwds):
+        self._savefun = savefun
+        self._kwds = kwds
+        self._flag = False
+
+    def __call__(self, filename, outdir, target):
+        self._process = multiprocessing.Process(
+            target=self.save,
+            args=(filename, outdir, target, self._savefun),
+            kwargs=self._kwds)
+        self._process.start()
+        self._flag = True
+
+    def finalize(self):
+        if self._flag:
+            self._process.join()
+
+
+class QueueWriter(Writer):
+    """Base class of queue snapshot writer.
+
+    This class a base class of writer that uses a queue. A Queue is created
+    when this class is constructed, and every time when `__call__` is
+    invoked, a snapshot task is put into the queue.
+
+    Args:
+        task: Callable object. Its `__call__` must have a same interface to
+            `Writer.__call__`. This object is put into the queue.
+    """
+
+    def __init__(self, savefun=util.save_npz, task=None):
+        raise NotImplementedError
+
+    def __call__(self, filename, outdir, target):
+        self._queue.put([self._task, filename, outdir, target])
+
+    def consume(self, queue):
+        while True:
+            task = queue.get()
+            if task is None:
+                queue.task_done()
+                return
+            else:
+                task[0](task[1], task[2], task[3])
+                queue.task_done()
+
+    def finalize(self):
+        self._queue.put(None)
+        self._queue.join()
+        self._consumer.join()
+
+
+class ThreadQueueWriter(QueueWriter):
+    """Writer that uses thread queue.
+
+    This class create a thread and a queue by `threading` and `Queue` module
+    respectively. The thread will be a consumer of this queue, and the main
+    thread will be a producer of this queue.
+
+    Args:
+        task: Callable object. Its `__call__` must have a same interface to
+            `Writer.__call__`. This object is put into the queue.
+    """
+
+    def __init__(self, savefun=util.save_npz, task=None):
+        if task is None:
+            self._task = SimpleWriter(savefun)
+        else:
+            self._task = task
+        self._queue = queue.Queue()
+        self._consumer = threading.Thread(target=self.consume,
+                                          args=(self._queue,))
+        self._consumer.start()
```

Comment:
It's not good to access base class'es private attributes (those starting with `_`).
How about making them public interface?
For example, adding
```
def create_queue(self):
    raise NotImplementedError()
```
in the base class and override it in the concrete classes.

And I think the construction logic can be implemented in the base class'es `QueueWriter.__init__()`, i.e.
```
if task is None:
    self._task = self.create_task(savefun)
else:
    self._task = task
self._queue = self.create_queue()
self._consumer = self.create_consumer()
self._consumer.start()
```